service/operator/service.go

```diff
@@ -126,83 +129,46 @@ func New(config Config) (*Service, error) {
 // Boot starts the service and implements the watch for azuretpr.
 func (s *Service) Boot() {
 	s.bootOnce.Do(func() {
-		err := s.tpr.CreateAndWait()
-		if tpr.IsAlreadyExists(err) {
-			s.Logger.Log("debug", "third party resource already exists")
-		} else if err != nil {
-			s.Logger.Log("error", fmt.Sprintf("%#v", err))
-			return
-		}
-
-		s.Logger.Log("debug", "starting list/watch")
+		o := func() error {
+			err := s.bootWithError()
+			if err != nil {
+				return microerror.Mask(err)
+			}
 
-		newResourceEventHandler := &cache.ResourceEventHandlerFuncs{
-			AddFunc:    s.addFunc,
-			DeleteFunc: s.deleteFunc,
-			UpdateFunc: s.updateFunc,
+			return nil
 		}
-		newZeroObjectFactory := &tpr.ZeroObjectFactoryFuncs{
-			NewObjectFunc:     func() runtime.Object { return &azuretpr.CustomObject{} },
-			NewObjectListFunc: func() runtime.Object { return &azuretpr.List{} },
+
+		n := func(err error, d time.Duration) {
+			s.logger.Log("warning", fmt.Sprintf("retrying operator boot due to error: %#v", microerror.Mask(err)))
 		}
 
-		s.tpr.NewInformer(newResourceEventHandler, newZeroObjectFactory).Run(nil)
+		err := backoff.RetryNotify(o, s.backoff, n)
+		if err != nil {
+			s.logger.Log("error", fmt.Sprintf("stop operator boot retries due to too many errors: %#v", microerror.Mask(err)))
+			os.Exit(1)
+		}
 	})
 }
 
-func (s *Service) addFunc(obj interface{}) {
-	// We lock the addFunc/deleteFunc/updateFunc to make sure only one
-	// addFunc/deleteFunc/updateFunc is executed at a time.
-	// addFunc/deleteFunc/updateFunc is not thread safe. This is important because
-	// the source of truth for the azure-operator are Azure resources.
-	// In case we would run the operator logic in parallel, we would run into race
-	// conditions.
-	s.mutex.Lock()
-	defer s.mutex.Unlock()
-
-	s.logger.Log("debug", "executing the operator's addFunc")
-
-	err := s.operatorFramework.ProcessCreate(obj, s.resources)
-	if err != nil {
-		s.logger.Log("error", fmt.Sprintf("%#v", err), "event", "create")
+func (s *Service) bootWithError() error {
+	err := s.tpr.CreateAndWait()
+	if tpr.IsAlreadyExists(err) {
+		s.Logger.Log("debug", "third party resource already exists")
+	} else if err != nil {
+		s.Logger.Log("error", fmt.Sprintf("%#v", err))
```

Comment:
When we handle the error here we do not need to log it. We would now log it twice. Lets get rid of this log though.

Reply:
Good point, done.